File: eos/signal-pdf.cc
/* vim: set sw=4 sts=4 et foldmethod=syntax : */

/*
 * Copyright (c) 2015-2026 Danny van Dyk
 * Copyright (c) 2019      Ahmet Kokulu
 *
 * This file is part of the EOS project. EOS is free software;
 * you can redistribute it and/or modify it under the terms of the GNU General
 * Public License version 2, as published by the Free Software Foundation.
 *
 * EOS is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <eos/b-decays/signal-pdfs.hh>
#include <eos/maths/power-of.hh>
#include <eos/rare-b-decays/signal-pdfs.hh>
#include <eos/signal-pdf-impl.hh>
#include <eos/signal-pdf.hh>
#include <eos/utils/concrete-signal-pdf.hh>
#include <eos/utils/density.hh>
#include <eos/utils/instantiation_policy-impl.hh>
#include <eos/utils/log.hh>
#include <eos/utils/private_implementation_pattern-impl.hh>
#include <eos/utils/wrapped_forward_iterator-impl.hh>

#include <algorithm>
#include <functional>
#include <map>
#include <ostream>

namespace eos
{
    namespace impl
    {
        std::map<QualifiedName, SignalPDFEntryPtr> signal_pdf_entries;
    }

    SignalPDFEntries::SignalPDFEntries() :
        _entries(&impl::signal_pdf_entries)
    {
        std::vector<std::function<SignalPDFSection()>> section_makers = {
            make_b_decays_pdf_section,
            make_rare_b_decays_pdf_section,
        };

        for (const auto & section_maker : section_makers)
        {
            for (const auto & group : section_maker())
            {
                _entries->insert(group.begin(), group.end());
            }
        }

        // add test entries to the list of available signal PDFs, but avoid adding it via a group/section
        // 1D Legendre PDF
        {
            using std::literals::string_literals::operator""s;

            auto name_and_entry_pair = make_signal_pdf("TestLegendre1D::P(z)",
                                                       "PDF for testing purpose only: 1D PDF based on Legendre polynomials as a function of the variable $z$.",
                                                       Options{},
                                                       "TestLegendre1D::UnnormalizedPDF(z)",
                                                       std::make_tuple("z"s),
                                                       "TestLegendre1D::NormalizationPDF(z)",
                                                       std::make_tuple("z_min"s, "z_max"s));

            _entries->insert(name_and_entry_pair);
        }
    }

    SignalPDFEntries::~SignalPDFEntries() = default;

    void
    SignalPDFEntries::insert_or_assign(const QualifiedName & key, const std::shared_ptr<const SignalPDFEntry> & value)
    {
        auto result = _entries->insert_or_assign(key, value);

        if (! result.second)
        {
            Log::instance()->message("[SignalPDFEntries.insert_or_assign]", ll_warning) << "Entry for signal PDF " << key.str() << " has been replaced.";
        }
    }

    SignalPDFNameError::SignalPDFNameError(const std::string & name) :
        Exception("SignalPDF name '" + name + "' is malformed")
    {
    }

    SignalPDFEntry::SignalPDFEntry() {}

    SignalPDFEntry::~SignalPDFEntry() = default;

    std::ostream &
    SignalPDFEntry::insert(std::ostream & os) const
    {
        os << "<empty SignalPDF description>" << std::endl;
        return os;
    }

    SignalPDFPtr
    SignalPDF::make(const QualifiedName & name, const Parameters & parameters, const Kinematics & kinematics, const Options & _options)
    {
        const std::map<QualifiedName, SignalPDFEntryPtr> & signal_pdf_entries = SignalPDFEntries::instance()->entries();

        // check if 'name' matches any of the implemented Signal PDFs
        {
            auto i = signal_pdf_entries.find(name);
            if (signal_pdf_entries.end() != i)
            {
                return i->second->make(parameters, kinematics, name.options() + _options);
            }
        }

        return SignalPDFPtr();
    }

    /* SignalPDFSections */

    class SignalPDFSections : public InstantiationPolicy<SignalPDFSections, Singleton>
    {
        private:
            std::vector<SignalPDFSection> _sections;

            SignalPDFSections()
            {
                // ensure that the observable entries have been generated already
                auto entries = std::distance(SignalPDFEntries::instance()->entries().begin(), SignalPDFEntries::instance()->entries().end());
                Log::instance()->message("SignalPDFSections::SignalPDFSections()", ll_debug) << "Total number of registered signal PDFs: " << entries;

                _sections = std::vector<SignalPDFSection>({
                    make_b_decays_pdf_section(),
                    make_rare_b_decays_pdf_section(),
                });
            }

            ~SignalPDFSections() = default;

        public:
            friend class InstantiationPolicy<SignalPDFSections, Singleton>;

            const std::vector<SignalPDFSection> &
            sections() const
            {
                return _sections;
            }
    };

    /* SignalPDFGroup */

    template <> struct WrappedForwardIteratorTraits<SignalPDFGroup::SignalPDFIteratorTag>
    {
            using UnderlyingIterator = std::map<QualifiedName, SignalPDFEntryPtr>::const_iterator;
    };
    template class WrappedForwardIterator<SignalPDFGroup::SignalPDFIteratorTag, const std::pair<const QualifiedName, SignalPDFEntryPtr>>;

    SignalPDFGroup::SignalPDFGroup(Implementation<SignalPDFGroup> * imp) :
        PrivateImplementationPattern<SignalPDFGroup>(imp)
    {
    }

    SignalPDFGroup::~SignalPDFGroup() = default;

    SignalPDFGroup::SignalPDFIterator
    SignalPDFGroup::begin() const
    {
        return _imp->entries.begin();
    }

    SignalPDFGroup::SignalPDFIterator
    SignalPDFGroup::end() const
    {
        return _imp->entries.end();
    }

    const std::string &
    SignalPDFGroup::name() const
    {
        return _imp->name;
    }

    const std::string &
    SignalPDFGroup::description() const
    {
        return _imp->description;
    }

    /* SignalPDFSection */

    template <> struct WrappedForwardIteratorTraits<SignalPDFSection::GroupIteratorTag>
    {
            using UnderlyingIterator = std::vector<SignalPDFGroup>::const_iterator;
    };
    template class WrappedForwardIterator<SignalPDFSection::GroupIteratorTag, const SignalPDFGroup &>;

    SignalPDFSection::SignalPDFSection(Implementation<SignalPDFSection> * imp) :
        PrivateImplementationPattern<SignalPDFSection>(imp)
    {
    }

    SignalPDFSection::~SignalPDFSection() = default;

    SignalPDFSection::GroupIterator
    SignalPDFSection::begin() const
    {
        return _imp->groups.begin();
    }

    SignalPDFSection::GroupIterator
    SignalPDFSection::end() const
    {
        return _imp->groups.end();
    }

    const std::string &
    SignalPDFSection::name() const
    {
        return _imp->name;
    }

    const std::string &
    SignalPDFSection::description() const
    {
        return _imp->description;
    }

    /* SignalPDFs */

    template <> struct WrappedForwardIteratorTraits<SignalPDFs::SignalPDFIteratorTag>
    {
            using UnderlyingIterator = std::map<QualifiedName, SignalPDFEntryPtr>::const_iterator;
    };
    template class WrappedForwardIterator<SignalPDFs::SignalPDFIteratorTag, const std::pair<const QualifiedName, SignalPDFEntryPtr>>;

    template <> struct WrappedForwardIteratorTraits<SignalPDFs::SectionIteratorTag>
    {
            using UnderlyingIterator = std::vector<SignalPDFSection>::const_iterator;
    };
    template class WrappedForwardIterator<SignalPDFs::SectionIteratorTag, const SignalPDFSection &>;

    template <> struct Implementation<SignalPDFs>
    {
            std::vector<SignalPDFSection> signal_pdf_sections;

            Implementation() :
                signal_pdf_sections(SignalPDFSections::instance()->sections())
            {
            }
    };

    SignalPDFs::SignalPDFs() :
        PrivateImplementationPattern<SignalPDFs>(new Implementation<SignalPDFs>())
    {
    }

    SignalPDFs::~SignalPDFs() {}

    SignalPDFEntryPtr
    SignalPDFs::operator[] (const QualifiedName & qn) const
    {
        const auto & signal_pdf_entries = SignalPDFEntries::instance()->entries();

        auto i = signal_pdf_entries.find(qn);
        if (i != signal_pdf_entries.end())
        {
            return i->second;
        }

        return SignalPDFEntryPtr(nullptr);
    }

    SignalPDFs::SignalPDFIterator
    SignalPDFs::begin() const
    {
        const auto & signal_pdf_entries = SignalPDFEntries::instance()->entries();

        return SignalPDFIterator(signal_pdf_entries.begin());
    }

    SignalPDFs::SignalPDFIterator
    SignalPDFs::end() const
    {
        const auto & signal_pdf_entries = SignalPDFEntries::instance()->entries();

        return SignalPDFIterator(signal_pdf_entries.end());
    }

    SignalPDFs::SectionIterator
    SignalPDFs::begin_sections() const
    {
        return SectionIterator(_imp->signal_pdf_sections.begin());
    }

    SignalPDFs::SectionIterator
    SignalPDFs::end_sections() const
    {
        return SectionIterator(_imp->signal_pdf_sections.end());
    }

    void
    SignalPDFs::insert(const QualifiedName & name, const std::string & description, const Options & options, const QualifiedName & numerator,
                       const std::vector<std::string> & numerator_kinematic_names, const QualifiedName & normalization,
                       const std::vector<std::string> & normalization_kinematic_names) const
    {
        const auto & observable_entries = ObservableEntries::instance()->entries();

        // the numerator and normalization must reference known observables; fail fast otherwise
        if (observable_entries.end() == observable_entries.find(numerator))
        {
            throw UnknownObservableError("Cannot create SignalPDF '" + name.str() + "': its numerator '" + numerator.str() + "' is not a known observable");
        }

        if (observable_entries.end() == observable_entries.find(normalization))
        {
            throw UnknownObservableError("Cannot create SignalPDF '" + name.str() + "': its normalization '" + normalization.str() + "' is not a known observable");
        }

        // each sampling variable 'v' (a numerator kinematic variable) should have matching bounds 'v_min' and 'v_max'
        // among the normalization kinematic variables; the Python sampling layer relies on this convention, so warn if it is not met
        for (const auto & variable : numerator_kinematic_names)
        {
            for (const auto & bound : { variable + "_min", variable + "_max" })
            {
                if (normalization_kinematic_names.end() == std::find(normalization_kinematic_names.begin(), normalization_kinematic_names.end(), bound))
                {
                    Log::instance()->message("[SignalPDFs.insert]", ll_warning)
                            << "SignalPDF '" << name.str() << "': the normalization is missing the bound '" << bound << "' for the sampling variable '" << variable
                            << "'; sampling from this PDF may not work as expected";
                }
            }
        }

        SignalPDFEntry * entry = new ConcreteSignalPDFEntry(name, description, options, numerator, normalization, numerator_kinematic_names, normalization_kinematic_names);

        SignalPDFEntries::instance()->insert_or_assign(name, std::shared_ptr<const SignalPDFEntry>(entry));
    }
} // namespace eos

File: python/eos/signal_pdf.py
# vim: set sw=4 sts=4 et tw=120 :

# Copyright (c) 2021-2023 Danny van Dyk
#
# This file is part of the EOS project. EOS is free software;
# you can redistribute it and/or modify it under the terms of the GNU General
# Public License version 2, as published by the Free Software Foundation.
#
# EOS is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
# Place, Suite 330, Boston, MA  02111-1307  USA

from _eos import _SignalPDF, _SignalPDFs
import eos
import numpy as np

class SignalPDF(_SignalPDF):
    def log_pdf(self, x):
        """
        Adapter for use with external optimization software (e.g. pypmc) to aid when sampling from the log(PDF).

        :param x: Phase space point, with the elements corresponding to the sorted list of variables in lexicographical order; inspect `self.variables`.
        :type x: iterable of float
        """
        for kv, v in zip(self.variables, x):
            kv.set(v)

        return self.evaluate()

    def sample_mcmc(self, N, stride, pre_N, preruns, cov_scale=0.1, start_point=None, rng=np.random.mtrand):
        """
        Return samples of the kinematic variables and the log(PDF).

        Obtains random samples of the log(PDF) using an adaptive Markov Chain Monte Carlo with PyPMC.
        A prerun with adaptations is carried out first and its samples are discarded.

        :param N: Number of samples that shall be returned
        :param stride: Stride, i.e., the number by which the actual amount of samples shall be thinned to return N samples.
        :param pre_N: Number of samples in each prerun.
        :param preruns: Number of preruns.
        :param cov_scale: Scale factor for the initial guess of the covariance matrix.
        :param start_point: Optional starting point for the chain
        :type start_point: list-like, optional
        :param rng: Optional random number generator (must be compatible with the requirements of pypmc.sampler.markov_chain.MarkovChain)

        :return: A tuple of the kinematic variables as array of size N and the log(PDF) as array of size N.

        .. note::
           This method requires the PyPMC python module, which can be installed from PyPI.
        """
        import pypmc
        try:
            from tqdm.auto import tqdm
            progressbar = tqdm
        except ImportError:
            progressbar = lambda x, **kw: x

        ind_lower = np.array([bound[0].evaluate() for bound in self.bounds])
        ind_upper = np.array([bound[1].evaluate() for bound in self.bounds])
        ind = pypmc.tools.indicator.hyperrectangle(ind_lower, ind_upper)

        log_target = pypmc.tools.indicator.merge_function_with_indicator(self.log_pdf, ind, -np.inf)

        # create initial covariance
        sigma = np.diag([np.square(bound[1].evaluate() - bound[0].evaluate()) / 12 * cov_scale for bound in self.bounds])
        log_proposal = pypmc.density.gauss.LocalGauss(sigma)

        # create start point, if not provided
        if start_point is None:
            u = np.array([rng.uniform(0.0, 1.0) for j in range(0, len(ind_lower))])
            ubar = 1.0 - u
            start_point = ubar * ind_upper + u * ind_lower

        # create MC sampler
        sampler = pypmc.sampler.markov_chain.AdaptiveMarkovChain(log_target, log_proposal, start_point, save_target_values=True, rng=rng)

        # pre run to adapt markov chains
        for i in progressbar(range(0, preruns), desc="Pre-runs", leave=False):
            eos.info(f'Prerun {i} out of {preruns}')
            accept_count = sampler.run(pre_N)
            accept_rate  = accept_count / pre_N * 100
            eos.info(f'Prerun {i}: acceptance rate is {accept_rate:3.0f}%')
            sampler.adapt()
        sampler.clear()

        # obtain final samples
        eos.info('Main run: started ...')
        sample_total  = N * stride
        sample_chunk  = sample_total // 100
        sample_chunks = [sample_chunk for i in range(0, 99)]
        sample_chunks.append(sample_total - 99 * sample_chunk)
        for current_chunk in progressbar(sample_chunks, desc="Main run", leave=False):
            accept_count = accept_count + sampler.run(current_chunk)
        accept_rate  = accept_count / (N * stride) * 100
        eos.info(f'Main run: acceptance rate is {accept_rate:3.0f}%')

        parameter_samples = sampler.samples[:][::stride]
        weights = sampler.target_values[:][::stride, 0]

        return(parameter_samples, weights)

    @staticmethod
    def make(name, parameters, kinematics, options):
        """
        Makes a new :class:`SignalPDF` object.

        :param name: The name of the probability density function (PDF). See `the complete list of PDFs <../reference/signal-pdfs.html>`_.
        :type name: eos.QualifiedName
        :param parameters: The set of parameters to which this PDF is bound.
        :type parameters: eos.Parameters
        :param kinematics: The set of kinematic variables to which this PDF is bound.
        :type kinematics: eos.Kinematics
        :param options: The set of options relevant to this PDF.
        :type options: eos.Options

        :return: The new PDF object.
        :rtype: eos.SignalPDF
        """
        pdf = _SignalPDF.make(name, parameters, kinematics, options)
        if pdf is None:
            return None

        pdf.__class__ = SignalPDF
        pdf.variables = list(map(
            lambda n: kinematics[n],
            filter(lambda n: not(n.endswith('_min') or n.endswith('_max')), [kv.name() for kv in kinematics])
        ))
        pdf.bounds = [(kinematics[v.name() + '_min'], kinematics[v.name() + '_max']) for v in pdf.variables]

        return pdf

class SignalPDFs(_SignalPDFs):
    """
    Represents the complete list of probability density functions (PDFs) known to EOS.

    Objects of this class are visualized as tables in Jupyter notebooks for easy
    overview. Filters can be applied through keyword arguments to the initialization.

    :param prefix: Only show observables whose qualified names contain the provided ``prefix`` in their prefix part.
    :type prefix: str
    :param name: Only show observables whose qualified names contain the provided ``name`` in their name part.
    :type name: str
    :param suffix: Only show observables whose qualified names contain the provided ``suffix`` in their suffix part.
    :type suffix: str

    See also `the complete list of signal PDFs <../reference/signal-pdfs.html>`_ in this documentation.
    """
    def __init__(self, prefix=None, name=None, suffix=None, showall=False):
        super().__init__()
        self.prefix=prefix
        self.name=name
        self.suffix=suffix
        self.showall=showall

    def filter_entry(self, qn):
        if self.prefix and not self.prefix in str(qn.prefix_part()):
            return False

        if self.name and not self.name in str(qn.name_part()):
            return False

        if self.suffix and not self.suffix in str(qn.suffix_part()):
            return False

        return True

    def _repr_html_(self):
        result = r'''
        <script>
            function toggle_group(group_title, id) {
                var table = group_title.parentNode.parentNode.parentNode.parentNode
                var query = 'tbody[id="' + id + '"]'
                var group = table.querySelector(query)
                if (group.style.visibility == "collapse") {
                    group.style.visibility = "visible"
                } else {
                    group.style.visibility = "collapse"
                }
            }
            function toggle_av(opt_anchor, id) {
                var query_dots   = 'span.dots[id="' + id + '"]'
                var query_values = 'span.values[id="' + id + '"]'
                var dots   = opt_anchor.querySelector(query_dots)
                var values = opt_anchor.querySelector(query_values)
                if (dots.style.display == "none") {
                    dots.style.display   = "inline"
                    values.style.display = "none"
                } else {
                    dots.style.display   = "none"
                    values.style.display = "inline"
                }
            }
        </script>
        <style>
            td.qn     { text-align: left;   }
            td.sym    { text-align: center; }
        </style>
        <table>
            <colgroup>
                <col width="100%" id="qn"         style="min-width: 300px; text-align: left">
            </colgroup>
            <thead>
                <tr>
                    <th>qualified name</th>
                </tr>
            </thead>
        '''
        group_id = 0
        observable_id = 0
        for section in self.sections():
            section_entries = 0
            section_result = fr'''
                <tr>
                    <th style="text-align:left" colspan=1><big>{section.name()}</big></th>
                </tr>'''
            for group in section:
                group_entries = 0
                group_result  = fr'''
                    <tbody>
                        <tr>
                            <th style="text-align:left" colspan=1>
                                <a style="text-decoration: none" onclick="toggle_group(this, 'grp{group_id}')">{group.name()}</a>
                            </th>
                        </tr>
                    </tbody>
                '''
                group_result += fr'''
                    <tbody style="visibility:collapse" id="grp{group_id}">
                    <tr>
                        <td style="text-align:left" colspan=1>{group.description()}</td>
                    </tr>
                '''
                for qn, entry in group:
                    if not self.filter_entry(qn):
                        continue

                    group_result += fr'''
                        <tr>
                            <th class="qn"     rowspan="1"><tt>{qn}</tt></th>
                        </tr>
                    '''

                    group_entries += 1
                    observable_id += 1

                group_result += '    </tbody>'
                group_id += 1
                section_entries += group_entries
                if group_entries > 0:
                    section_result += group_result
            if section_entries > 0:
                result += section_result
        result += r'</table>'

        return(result)
